syntax = "proto3";

package credit;

import "google/protobuf/empty.proto";

service CreditService {
  rpc CreateCreditProfile (CreateCreditProfileRequest) returns (CreditProfileResponse) {}
  rpc GetAllCreditProfiles (google.protobuf.Empty) returns (GetAllCreditProfilesResponse) {}
  rpc CreateDefaultLoanApplication (CreateDefaultLoanApplicationRequest) returns (CreateDefaultLoanApplicationResponse) {}
  rpc GetLoanOptions (GetLoanOptionsRequest) returns (GetLoanOptionsResponse) {}
  rpc UpdateCheckoutSessionForLoan (UpdateCheckoutSessionForLoanRequest) returns (UpdateCheckoutSessionForLoanResponse) {}
  rpc SelectLoan (SelectLoanRequest) returns (SelectLoanResponse) {}
  rpc GetLoanForCheckoutSession (GetLoanForCheckoutSessionRequest) returns (GetLoanForCheckoutSessionResponse) {}
}

message CreateCreditProfileRequest {
  string name = 1;
  int32 credit_score = 2;
  int32 number_of_accounts = 3;
  float credit_utilization_ratio = 4;
  int32 recent_soft_inquiries = 5;
  int32 bankruptcies = 6;
  int32 tax_liens = 7;
  int32 judgments = 8;
}

message CreditProfileResponse {
  string id = 1;
  string user_id = 2;
  string name = 3;
  int32 credit_score = 4;
  int32 number_of_accounts = 5;
  float credit_utilization_ratio = 6;
  int32 recent_soft_inquiries = 7;
  int32 bankruptcies = 8;
  int32 tax_liens = 9;
  int32 judgments = 10;
  string created_at = 11;
  string updated_at = 12;
}

message GetAllCreditProfilesResponse {
  repeated CreditProfileResponse profiles = 1;
}

message CreateDefaultLoanApplicationRequest {
  int64 loan_amount_cents = 1;
  int32 merchant_id = 2;
}

message CreateDefaultLoanApplicationResponse {
  bool success = 1;
  string loan_id = 2;
}

message GetLoanOptionsRequest {
  string user_id = 1;
  string session_id = 2;
}

message GetLoanOptionsResponse {
  repeated LoanOption loan_options = 1;
}

message LoanOption {
  string id = 1;
  int32 loan_amount_cents = 2;
  int32 loan_term_months = 3;
  float interest_rate = 4;
  float monthly_payment_cents = 5;
  float total_payment_amount_cents = 6;
}

message SelectLoanRequest {
  string user_id = 1;
  string checkout_session_id = 2;
  int32 loan_term_months = 3;
  float interest_rate = 4;
  int32 monthly_payment_cents = 5;
  int32 total_payment_amount_cents = 6;
}

message UpdateCheckoutSessionForLoanRequest {
  string loan_id = 1;
  string checkout_session_id = 2;
}

message UpdateCheckoutSessionForLoanResponse {
  bool success = 1;
}

message SelectLoanResponse {
  bool success = 1;
}

message GetLoanForCheckoutSessionRequest {
  string checkout_session_id = 1;
  string user_id = 2;
}

message GetLoanForCheckoutSessionResponse {
  string status = 1;
}
